### PR TITLE
Generalize artifact finding to support multiple artifact groups

### PR DESCRIPTION
**Pull request by ScottTodd**
## Motivation

As part of testing https://github.com/ROCm/TheRock/pull/3608 for https://github.com/ROCm/TheRock/issues/3330, I wanted a scriptable way to find recent artifacts that include multiple artifact groups, not just one.

This will also help with https://github.com/ROCm/TheRock/issues/3399, where multi-arch CI will want a way to find a single baseline commit containing artifacts for all artifact groups that it plans on building.

## Technical Details

I've coded defensively against a few edge cases (e.g. a more recent commit including artifacts for one group but not another), but I might have missed a few.

## Test Plan

* New and existing unit tests
* Run the scripts locally
* Skipping CI since these scripts are not yet used directly or indirectly by any workflows

## Test Results

```
(.venv) λ python build_tools/find_latest_artifacts.py --artifact-group gfx110X-all gfx120X-all -v
Searching 50 commits on ROCm/TheRock/main for 2 group(s): gfx110X-all, gfx120X-all...
  [1/50] Checking 7a38803c...
Retrieving bucket info...
  (explicit) github_repository: ROCm/TheRock
  workflow_run_id             : 22417254779
  head_github_repository      : ROCm/TheRock
  is_pr_from_fork             : False
Retrieved bucket info:
  external_repo:
  bucket       : therock-ci-artifacts
    No artifacts found
  [2/50] Checking 06457686...
Retrieving bucket info...
  (explicit) github_repository: ROCm/TheRock
  workflow_run_id             : 22407066489
  head_github_repository      : ROCm/TheRock
  is_pr_from_fork             : False
Retrieved bucket info:
  external_repo:
  bucket       : therock-ci-artifacts
    Found all 2 group(s): run(s) 22407066489
Artifact info:
  Git repository:      ROCm/TheRock
  Git commit:          064576864e0951e5398e0b6a5dcc03efb80e9ce3
  Git commit URL:      https://github.com/ROCm/TheRock/commit/064576864e0951e5398e0b6a5dcc03efb80e9ce3
  Platform:            windows
  Artifact group:      gfx110X-all
  Workflow name:       ci.yml
  Workflow run ID:     22407066489
  Workflow run URL:    https://github.com/ROCm/TheRock/actions/runs/22407066489
  Workflow run status: completed (failure)
  S3 Bucket:           therock-ci-artifacts
  S3 Path:             22407066489-windows/
  S3 Index:            https://therock-ci-artifacts.s3.amazonaws.com/22407066489-windows/index-gfx110X-all.html

Artifact info:
  Git repository:      ROCm/TheRock
  Git commit:          064576864e0951e5398e0b6a5dcc03efb80e9ce3
  Git commit URL:      https://github.com/ROCm/TheRock/commit/064576864e0951e5398e0b6a5dcc03efb80e9ce3
  Platform:            windows
  Artifact group:      gfx120X-all
  Workflow name:       ci.yml
  Workflow run ID:     22407066489
  Workflow run URL:    https://github.com/ROCm/TheRock/actions/runs/22407066489
  Workflow run status: completed (failure)
  S3 Bucket:           therock-ci-artifacts
  S3 Path:             22407066489-windows/
  S3 Index:            https://therock-ci-artifacts.s3.amazonaws.com/22407066489-windows/index-gfx120X-all.html
```

That found:
* Commit: https://github.com/ROCm/TheRock/commit/064576864e0951e5398e0b6a5dcc03efb80e9ce3
* Workflow run: https://github.com/ROCm/TheRock/actions/runs/22407066489
* gfx110X artifacts: https://therock-ci-artifacts.s3.amazonaws.com/22407066489-windows/index-gfx110X-all.html
* gfx120X artifacts: https://therock-ci-artifacts.s3.amazonaws.com/22407066489-windows/index-gfx120X-all.html

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

### CODE REVIEW

File: build_tools/find_artifacts_for_commit.py [hunk 1]
```diff
@@ -152,59 +122,71 @@ def check_if_artifacts_exist(info: ArtifactRunInfo) -> bool:
...
+            github_repository=github_repository_name,
             workflow_run=workflow_run,
         )
-
-        if check_if_artifacts_exist(info):
-            return info
-
-    # No runs had artifacts available
-    return None
+        for group in artifact_groups:
```
Comment (geomin12) [inline build_tools/find_artifacts_for_commit.py:169]: just making sure, are the runs ordered correctly based on newer / older? could we maybe add a sort by time to guarantee this?

I see the comment `(most recent first)` and i would assume API returns in order, but seeing a code guarantee may ease my concern

Comment (ScottTodd) [reply build_tools/find_artifacts_for_commit.py:169]: I thought it would be safe to trust the API here, but https://github.com/orgs/community/discussions/157545 suggests that sorting may not always be purely chronological, it may also favor successful runs.

I've added some explicit sorting in both the API request and a client-side sort, along with a unit test showing the client side behavior and documenting the created_at timestamp format.
